src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+        }
+    }
+}
+
+
+impl RouteService {
+
+    /// Creates a new route service
+    pub fn new() -> RouteService {
+        RouteService {
+            history: window().history(),
+            location: window().location().unwrap(),
+            event_listener: None,
+            callback: None
+        }
+    }
+
+
+    /// Creates the callback used in the main routing logic.
+    ///
+    /// The callback takes a string, parses it into a url, and then uses the result of that
+    /// to create a message that the component will use to update itself with.
+    // TODO I would like to include this in register_router
+    pub fn create_routing_callback<COMP, CTX>(context: &mut Env<CTX, COMP>) -> Callback<RouteResult>
+        where
+            COMP: Component<CTX>,
+            COMP::Msg: From<RouteResult>,
+            CTX: 'static
+    {
+        return context.send_back(|route_result: RouteResult| {
+            println!("Callback path changed {:?}", route_result); // Remove me
+            COMP::Msg::from(route_result)
+        })
+    }
+
+    /// Will return the current route info based on the location API.
+    // TODO this should probably return a RouteResult and avoid expecting
+    pub fn get_current_route_info(&mut self) -> RouteInfo {
+        // If the location api errors, recover by redirecting to a valid address
+        let href = self.get_location().expect("Couldn't get href from location Api");
+        let url = Url::parse(&href).expect("The href returned from the location api should always be parsable.");
+        RouteInfo::from(url)
+    }
+
+    /// Registers the router.
+    /// There can only be one router.
+    /// The component in which it is set up will be the source from which routing logic emanates.
+    pub fn register_router(&mut self, callback: Callback<RouteResult>)
+    {
+        if let Some(_) = self.event_listener {
+            panic!("You cannot register two separate routers.");
+        }
+
+        // Hold on to the callback so it can be used to update the main router component
+        // when a user clicks a link, independent of the event listener.
+        self.callback = Some(callback.clone());
+
+        // Set the event listener to listen for the history's pop state events and call the callback when that occurs
+        self.event_listener = Some( window().add_event_listener(move |event: PopStateEvent| {
+            let state_value: Value = event.state();
+
+            if let Ok(state) = String::try_from(state_value) {
+                callback.emit(RouteInfo::parse(&state))
+            } else {
+                eprintln!("Nothing farther back in history, not calling routing callback.");
+            }
+        }));
+    }
+
+
+    /// Sets the route via the history api.
+    /// This does not by itself make any changes to Yew's state.
+    fn set_route(&mut self, route_info: RouteInfo) -> RouteDidChange {
+        if route_info != self.get_current_route_info() {
```

Comment:
the spec actually allows[ you to do this](https://developer.mozilla.org/en-US/docs/Web/API/History_API#The_pushState()_method), mainly to update the state (I think)

> The new URL can be any URL in the same origin as the current URL.

Reply:
~~I store the current route string as the state. I don't know if this is standard practice, or if in other routers they store the route as well as (part of?) the model. Conceptually you could restore the routed-to model's data on a back event using a tuple of `(String, Option<Model>)` where `Model` implements `Storable`, I decided pretty early on that I only wanted to work with Strings to keep my implementation simple.~~ Pre-Edit: I wrote the above text before I came to the conclusion that I actually want the ability to store arbitrary state in the History API. It no longer applies.

On to addressing your comment:

I specifically did not want it to be possible to add more history entries to the same page you are already on. Just like how a plain HTML page won't reload the page when you click a link that points to that same page, I don't want webapps built using Yew to be able to add history entries by clicking a link that appears in a header 30 times in a row. If they did that without that conditional, they would have to press the browser's back button 30 times to go to a prior page on the site. 

The principle of least surprise indicates that Yew should prevent consecutive duplicate entries to the History API.